### ISSUE TITLE
libselinux: expose library version in headers or API

### ISSUE THREAD

**Issue by koutheir** [SELinuxProject/selinux#296]
Please provide a means to discover the version of `libselinux` either from the header files, or (preferably) from the compiled library. A version encoded on one 32-bit word as follows seems reasonable:
```c
uint32_t selinux_version() {
    return (MAJOR << 16) | MINOR;
}
```

This would allow clients to choose which APIs they can call (and which flags they can use) depending on the version of `libselinux` currently installed on the system.

**Comment by fishilico** [comment]
Hello,
This version is exposed in pkg-config file. For example in Debian sid:

```console
$ sudo apt install libselinux1-dev
$ cat /usr/lib/x86_64-linux-gnu/pkgconfig/libselinux.pc
prefix=/usr
exec_prefix=${prefix}
libdir=/usr/lib
includedir=/usr/include

Name: libselinux
Description: SELinux utility library
Version: 3.1
URL: http://userspace.selinuxproject.org/
Requires.private: libsepol libpcre2-8
Libs: -L${libdir} -lselinux
Cflags: -I${includedir}
```
Here, `Version: 3.1` gives the package version.

From the compiled library, you can enumerate the exported symbols (using `nm -D`) to list the symbols with their API versions. Right now, all symbols use `LIBSELINUX_1.0` (cf. https://github.com/SELinuxProject/selinux/blob/cedcbaa8240ce5c79468ee3d68cfe7fe6735bc1a/libselinux/src/libselinux.map) and this can evolve over time.

**Comment by koutheir** [comment]
>  This version is exposed in pkg-config file. For example in Debian sid:

This versioning is independent of the library sources and binaries. It is at the package level, so it depends on the package management system. The information I'm asking for should be independent of the packaging system.

> From the compiled library, you can enumerate the exported symbols (using nm -D) to list the symbols with their API versions. Right now, all symbols use LIBSELINUX_1.0 (cf. https://github.com/SELinuxProject/selinux/blob/cedcbaa8240ce5c79468ee3d68cfe7fe6735bc1a/libselinux/src/libselinux.map) and this can evolve over time.

This has nothing to do with the library version. This is the version of the symbol itself. This simply tracks ABI compatibility of the affected symbols and allows the same symbol name to be provided with different implementations in the same library version.

**Comment by fishilico** [comment]
> This versioning is independent of the library sources and binaries. It is at the package level, so it depends on the package management system. The information I'm asking for should be independent of the packaging system.

No, this is the version of the library and it directly comes from libselinux. More precisely it is the version from the `VERSION` file of https://github.com/SELinuxProject/selinux/tree/cedcbaa8240ce5c79468ee3d68cfe7fe6735bc1a/libselinux, populated in `libselinux.pc` using a `sed` command in libselinux's `Makefile`: https://github.com/SELinuxProject/selinux/blob/master/libselinux/src/Makefile#L160-L161

Also, in my humble opinion pkgconfig files can hardly be considered a "packaging system": they are mostly used by build systems to resolve build-time dependencies, to provide information such as the library version so that projects (using autoconf, cmake, ...) can check this version. For information, pkgconfig files can also be used from more complex build systems such as Rust's one, using for example https://crates.io/crates/pkg-config.

> This has nothing to do with the library version. This is the version of the symbol itself. This simply tracks ABI compatibility of the affected symbols and allows the same symbol name to be provided with different implementations in the same library version.

Yes and it answers your need "This would allow clients to choose which APIs they can call (and which flags they can use)", which is why I indicated it.

It seems that I do not understand your use-case. Could you please describe more precisely what you are trying to do and why you need a `selinux_version` function?

**Comment by koutheir** [comment]
> Also, in my humble opinion pkgconfig files can hardly be considered a "packaging system": they are mostly used by build systems to resolve build-time dependencies, to provide information such as the library version so that projects (using autoconf, cmake, ...) can check this version. For information, pkgconfig files can also be used from more complex build systems such as Rust's one, using for example https://crates.io/crates/pkg-config.

A client program (in binary form) depending on the presence of some version of libselinux does not assume the presence of pkgconfig or any other packaging system.Yet it needs to find out which version of the library is installed.

> Yes and it answers your need "This would allow clients to choose which APIs they can call (and which flags they can use)", which is why I indicated it.

Currently, it doesn't, because nobody incremented the `LIBSELINUX_1.0` version label when new flags became accepted by some libselinux APIs, so a client calling that API cannot know if it's legal for it to specify that flag as an argument, or fall back to something else.

> It seems that I do not understand your use-case. Could you please describe more precisely what you are trying to do and why you need a selinux_version function?

I already specified my use case:
> "This would allow clients to choose which APIs they can call (and which flags they can use) depending on the version of libselinux currently installed on the system."

**Comment by bachradsusi** [comment]
I'd say that libselinux API is quite stable. OTOH libsepol provides symbols with 3 different versions - https://github.com/SELinuxProject/selinux/blob/master/libsepol/src/libsepol.map.in, libsemanage has 2 versions https://github.com/SELinuxProject/selinux/blob/master/libsemanage/src/libsemanage.map

Since 3.0 release,  abi/api compatibility is checked before new release and if there's a change, affected symbol versions are incremented. If you know about something missed please let us know.

**Comment by koutheir** [comment]
> I'd say that libselinux API is quite stable...

I'm currently focused on `libselinux`, but this situation might also apply to other libraries. I don't currently have enough experience with the other libraries.

> Since 3.0 release...

That's good news. My use case covers `libselinux` version 2.8 and later too.

>  ..., abi/api compatibility is checked before new release and if there's a change, affected symbol versions are incremented.

I would like to notice that an existing API that accepts new flags without changing its prototype is a change I'm interested in. Such a change can be considered ABI-breaking, depending on how the old implementation behaves when unrecognized flags are given as input. If unrecognized flags cause an error, then such a change breaks the ABI.

If you think ABI compatibility is covered completely using ELF versioning, then you can disregard this issue. Given my own experience, strict checking of ABI compatibility is pretty difficult to perform, so providing other sources of information about which version of the library the caller is dealing with allows graceful handling of corners missed by ABI checks.

**Comment by bachradsusi** [comment]
https://plautrba.fedorapeople.org/libselinux-2.8-master-compat-report.html

This is compat report between 2.8 and current master generated by abi-compliance-checker

**Comment by koutheir** [comment]
The `abi-compliance-checker` is a good tool, but it doesn't detect all ABI issues. I think an example might illustrate better my concerns:

Between `libselinux` versions `2.9` and `3.1`:
- Two new flags `SELINUX_RESTORECON_SKIP_DIGEST` and `SELINUX_RESTORECON_CONFLICT_ERROR` were added to the flags accepted by `selinux_restorecon()`.
- `selinux_restorecon()` already accepted an existing set of flags: `SELINUX_RESTORECON_*`.
- `selinux_restorecon()` has only one ELF version: `LIBSELINUX_1.0`.
- The signature of `selinux_restorecon()` did **not** change when the new flags were introduced.

The question now is: how does `selinux_restorecon()` behave if unrecognized flags are given to it?
1. If `selinux_restorecon()` ignores unknown flags.
2. If `selinux_restorecon()` returns an error when unknown flags are given.

Case (1) is the nice behavior. A caller can simply provide the new flags to any implementation of `selinux_restorecon@LIBSELINUX_1.0` and expect it to be compatible.

In case (2), the addition of new accepted flags breaks the ABI. A caller that specifically calls `selinux_restorecon@LIBSELINUX_1.0` and provides the new flags might be surprised by an unexpected error. A potential solution based on ELF versioning would be to mark the implementation of `selinux_restorecon()` accepting the new flags `selinux_restorecon@LIBSELINUX_1.1` and `selinux_restorecon@LIBSELINUX_1.2` (because in this example, each new flag was introduced in a separate version). This would be a reasonable solution to this issue, but I highly doubt `abi-compliance-checker` can detect this ABI compliance problem.

Because such an issue is currently not solved through ELF versioning, I suggested yet another piece of information that allows detecting this corner case.

**Comment by bachradsusi** [comment]
In this particular case, unknown flags are ignored, see https://github.com/SELinuxProject/selinux/blob/f25d03d75d3504e71b1fcb7cee74c0d2cba1cda0/libselinux/src/selinux_restorecon.c#L804 - this is `selinux_restorecon()` before  `SELINUX_RESTORECON_CONFLICT_ERROR` is added.

But if you don't depend on behavior related to `SELINUX_RESTORECON_CONFLICT_ERROR` or `SELINUX_RESTORECON_SKIP_DIGEST` why would you use these flags?

**Comment by koutheir** [comment]
> In this particular case, unknown flags are ignored

Good, so this example hits *Case (1)*. Should I assume that all cases of this nature also behave similarly?

> But if you don't depend on behavior related to SELINUX_RESTORECON_CONFLICT_ERROR or SELINUX_RESTORECON_SKIP_DIGEST why would you use these flags?

I'm not using these flags. My particular use case is making a high-level library that allows its clients to use these flags, and maybe providing fallback in case the flags cannot be used. This is only an example to illustrate a reason I created this issue.

**Comment by koutheir** [comment]
@stephensmalley, was this issue closed because the feature was implemented? If so, can you please tell me which API implements it?

**Comment by stephensmalley** [comment]
No, we have no plans for implementing this and no concrete use case where it would make a difference.

**Comment by koutheir** [comment]
Thanks for marking this as not planned. That clarifies the situation.
